Implement drag start from site isolated iframes

achristensen07 opened the pull request:
#### f309c92390b8a23152607f8085a692175802ffca
<pre>
Implement drag start from site isolated iframes
<a href="https://bugs.webkit.org/show_bug.cgi?id=292647">https://bugs.webkit.org/show_bug.cgi?id=292647</a>
<a href="https://rdar.apple.com/150813333">rdar://150813333</a>

Reviewed by Wenson Hsieh.

* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::tryToBeginDragAtPoint):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _dragInteraction:itemsForAddingToSession:withTouchAtPoint:completion:]):
(-[WKContentView _dragInteraction:prepareForSession:completion:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::requestDragStart):
(WebKit::WebPageProxy::requestAdditionalItemsForDragSession):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDragStart):
(WebKit::WebPage::requestAdditionalItemsForDragSession):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _requestStartDataInteraction:globalPosition:]):

Canonical link: <a href="https://commits.webkit.org/294753@main">https://commits.webkit.org/294753@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6fd3f732af44f6b45b8106fd762e06c2540ead40

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102539 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22208 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12522 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107700 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53176 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104578 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22516 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30714 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78003 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34993 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105545 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17444 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92555 "Found 1 new API test failure: TestWebKitAPI.WebKit.PrintFrame (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58339 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17282 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10583 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52533 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87097 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10652 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110075 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29671 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21881 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86982 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30035 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88749 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86580 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31409 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9138 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23921 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16702 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29599 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34911 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29410 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32733 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30969 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -1111,17 +1112,27 @@ static void dispatchSyntheticMouseMove(LocalFrame& mainFrame, const WebCore::Flo
+        co_return { false };
 
-    localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition, WTFMove(completionHandler));
+    auto handledOrTransformer = co_await AwaitableFromCompletionHandler<Expected<bool, RemoteFrameGeometryTransformer>> { [=] (auto completionHandler) {
+        localRootFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition, WTFMove(completionHandler));
+    } };
+    if (handledOrTransformer)
+        co_return { *handledOrTransformer };
+    auto& transformer = handledOrTransformer.error();
+    co_return { DragInitiationResult::RemoteFrameData {
+        transformer.remoteFrameID(),
+        transformer.transformToRemoteFrameCoordinates(clientPosition),
+        transformer.transformToRemoteFrameCoordinates(globalPosition)
+    } };
```
Is there any way we could share some of this code with `requestAdditionalItemsForDragSession` below?

achristensen07 replied:
We could make a shared function that takes a transformer, client position, and global position, and returns a RemoteFrameData.  I don't think that would be better than this.  It wouldn't reduce duplicate code, and it wouldn't really express significant shared functionality.

whsieh replied:
> It wouldn't reduce duplicate code, and it wouldn't really express significant shared functionality.

Hm.. not so sure about this. It would mean we could consolidate this logic:

```
    auto& transformer = handledOrTransformer.error();
    co_return { DragInitiationResult::RemoteFrameData {
        transformer.remoteFrameID(),
        transformer.transformToRemoteFrameCoordinates(clientPosition),
        transformer.transformToRemoteFrameCoordinates(globalPosition)
    } };
```

One could imagine just adding a helper to the transformer itself, like `computeRemoteFrameData`. But what you have here seems fine as well, since it's only repeated once.

whsieh replied:
Perhaps more importantly — I just noticed the RAII state above:

```
SetForScope allowedActionsForScope(m_allowedDragSourceActions, allowedActionsMask);
```

…from a quick scan of the code, it seems that's consulted through `DragController` when starting the drag:

```
OptionSet<DragSourceAction> DragController::delegateDragSourceAction(const IntPoint& rootViewPoint)
{
    m_dragSourceAction = client().dragSourceActionMaskForPoint(rootViewPoint);
    return m_dragSourceAction;
}
```

Could you verify that this is safe? (Do we need to set the RAII flags again when we call into `tryToBeginDragAtPoint`?)

(Edit: the code in `requestAdditionalItemsForDragSession` potentially has the same issue)

achristensen07 replied:
The SetForScope looks safe and correct to me.  The scope changed if tryToBeginDragAtPoint doesn't return immediately, which isn't the case currently, but even when we introduce that it looks correct and safe.

If we unified the RemoteFrameData constructor call, it would be a function that takes 3 parameters and basically just passes them to a constructor.  In my opinion that isn't meaningful code sharing, but it's definitely a matter of opinion.

achristensen07 committed:
Implement drag start from site isolated iframes
https://bugs.webkit.org/show_bug.cgi?id=292647
rdar://150813333

Reviewed by Wenson Hsieh.

* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::tryToBeginDragAtPoint):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _dragInteraction:itemsForAddingToSession:withTouchAtPoint:completion:]):
(-[WKContentView _dragInteraction:prepareForSession:completion:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::requestDragStart):
(WebKit::WebPageProxy::requestAdditionalItemsForDragSession):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDragStart):
(WebKit::WebPage::requestAdditionalItemsForDragSession):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _requestStartDataInteraction:globalPosition:]):

Canonical link: https://commits.webkit.org/294753@main
